File: problems/tetris/starter/java/Piece.java
public class Piece {
    // The location of the origin (top left) of the matrix representing the peice.
    public int x, y;

    // The rotation, from 0-3 going clockwise from the starting position.
    public int rotation;

    // The matrix representing the piece in its current orientation.
    // This matrix can be a different size depending on the piece.
    public boolean[][] matrix;

    // The keycode of the piece.
    public char type;

    // Construct a piece from the key, aka IJLOSTZ.
    public Piece(char key) {
        type = key;
        switch (key) {
            case 'I': matrix = PIECES.I;
                      break;
            case 'J': matrix = PIECES.J;
                      break;
            case 'L': matrix = PIECES.L;
                      break;
            case 'O': matrix = PIECES.O;
                      break;
            case 'S': matrix = PIECES.S;
                      break;
            case 'Z': matrix = PIECES.Z;
                      break;
            case 'T': matrix = PIECES.T;
                      break;
            default : throw new RuntimeException("Invalid piece initialization character.");
        } rotation = x = y = 0; // intialize piece to orientation 0 at top left
    }

    // Rotate the piece 90 degrees clockwise.
    public void rotate() {
        final boolean[][] ret = new boolean[matrix.length][matrix.length];
        for (int x=0; x<ret.length; x++) for (int y=0; y<ret.length; y++) {
            ret[y][ret.length-1-x] = matrix[x][y];
        } matrix = ret; rotation += 1;
        if (rotation > 3) rotation -= 4;
    }

    // Set the piece to a specified orientation, 0-3 inclusive going clockwise from start.
    public void setRotation(int rotation) {
        if (rotation < 0 || rotation > 3)
            throw new RuntimeException("Rotation must be between 0 and 3 inclusive.");
        while (rotation != this.rotation) rotate();
    }

    // Templates for pieces IJLOSZT in their starting orientations.
    public static class PIECES {
        public static final boolean[][] I = {
            { false, false, false, false },
            { true,  true,  true,  true },
            { false, false, false, false },
            { false, false, false, false }
        };

        public static final boolean[][] J = {
            { true,  false, false },
            { true,  true,  true  },
            { false, false, false },
        };

        public static final boolean[][] L = {
            { false, false, true  },
            { true,  true,  true  },
            { false, false, false },
        };

        public static final boolean[][] O = {
            { true, true },
            { true, true }
        };

        public static final boolean[][] S = {
            { false, true,  true  },
            { true,  true,  false },
            { false, false, false },
        };

        public static final boolean[][] Z = {
            { true,  true,  false },
            { false, true,  true  },
            { false, false, false },
        };

        public static final boolean[][] T = {
            { false, true,  false },
            { true,  true,  true  },
            { false, false, false }
        };
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();
        for (boolean[] row : matrix) {
            for (boolean is : row) builder.append(is?'X':'O');
            builder.append('\n');
        } return builder.toString();
    }
}

File: problems/tetris/starter/python/Piece.py
import copy

class Piece:
    def __init__(self, key):
        # The location of the origin (top left) of the matrix representing the peice.
        self.x = 0
        self.y = 0

        # The rotation, from 0-3 going clockwise from the starting position.
        self.rotation = 0

        if key in {'I', 'J', 'L', 'O', 'S', 'Z', 'T'}:
            self.matrix = copy.deepcopy(getattr(self,key))
            self.type = key
        else:
            raise RuntimeError('Invalid piece initialization character.')

    # Rotate the piece 90 degrees clockwise.
    def rotate(self):
        ret = copy.deepcopy(self.matrix)
        for x in range(0,len(ret)):
            for y in range (0,len(ret)):
                ret[y][len(ret)-1-x] = self.matrix[x][y];
        self.matrix = ret
        self.rotation += 1
        if self.rotation > 3: self.rotation -= 4

    # Set the piece to a specified orientation, 0-3 inclusive going clockwise from start.
    def setRotation(self, rotation):
        if self.rotation < 0 or self.rotation > 3:
            raise RuntimeError('Rotation must be between 0 and 3 inclusive')
        while self.rotation != rotation:
            self.rotate()


    # Templates for pieces IJLOSZT in their starting orientations.
    I = [[ False, False, False, False ],
         [ True,  True,  True,  True ],
         [ False, False, False, False ],
         [ False, False, False, False ]
    ]

    J = [[ True,  False, False ],
         [ True,  True,  True  ],
         [ False, False, False ],
    ]

    L = [[ False, False, True  ],
         [ True,  True,  True  ],
         [ False, False, False ],
    ]

    O = [[ True, True ],
         [ True, True ]
    ]

    S = [[ False, True,  True  ],
         [ True,  True,  False ],
         [ False, False, False ],
    ]

    Z = [[ True,  True,  False ],
         [ False, True,  True  ],
         [ False, False, False ],
    ]

    T = [[ False, True,  False ],
         [ True,  True,  True  ],
         [ False, False, False ]
    ]

    def __str__(self):
        outstring = ''
        for row in self.matrix:
            for val in row:
                outstring += 'X' if val else 'O'
            outstring += '\n'
        return outstring
